File: frontend/src/utils/passwordValidator.ts
/**
 * Password validation utility matching backend requirements
 */

export interface PasswordValidationResult {
    valid: boolean;
    errors: string[];
    strength: 'weak' | 'medium' | 'strong';
}

export function validatePassword(password: string): PasswordValidationResult {
    const errors: string[] = [];

    if (!password || password.length < 8) {
        errors.push('Password must be at least 8 characters long');
    }

    if (!/[A-Z]/.test(password)) {
        errors.push('Must contain at least one uppercase letter');
    }

    if (!/[a-z]/.test(password)) {
        errors.push('Must contain at least one lowercase letter');
    }

    if (!/[0-9]/.test(password)) {
        errors.push('Must contain at least one number');
    }

    if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
        errors.push('Must contain at least one special character');
    }

    // Determine strength
    let strength: 'weak' | 'medium' | 'strong' = 'weak';
    if (errors.length === 0) {
        if (password.length >= 12 && /[A-Z]/.test(password) && /[a-z]/.test(password) &&
            /[0-9]/.test(password) && /[!@#$%^&*(),.?":{}|<>]/.test(password)) {
            strength = 'strong';
        } else {
            strength = 'medium';
        }
    }

    return {
        valid: errors.length === 0,
        errors,
        strength
    };
}

File: backend/src/main/java/com/m4hub/backend/util/PasswordValidator.java
package com.m4hub.backend.util;

import org.springframework.stereotype.Component;

import java.util.regex.Pattern;

/**
 * Password validation utility
 * Enforces strong password requirements across the application
 */
@Component
public class PasswordValidator {

    private static final int MIN_LENGTH = 8;
    private static final Pattern UPPERCASE_PATTERN = Pattern.compile("[A-Z]");
    private static final Pattern LOWERCASE_PATTERN = Pattern.compile("[a-z]");
    private static final Pattern DIGIT_PATTERN = Pattern.compile("[0-9]");
    private static final Pattern SPECIAL_CHAR_PATTERN = Pattern.compile("[!@#$%^&*(),.?\":{}|<>]");

    /**
     * Validates password against all security requirements
     * 
     * @param password The password to validate
     * @return ValidationResult with success status and error message if invalid
     */
    public ValidationResult validate(String password) {
        if (password == null || password.isEmpty()) {
            return new ValidationResult(false, "Password is required");
        }

        if (password.length() < MIN_LENGTH) {
            return new ValidationResult(false,
                    String.format("Password must be at least %d characters long", MIN_LENGTH));
        }

        if (!UPPERCASE_PATTERN.matcher(password).find()) {
            return new ValidationResult(false,
                    "Password must contain at least one uppercase letter");
        }

        if (!LOWERCASE_PATTERN.matcher(password).find()) {
            return new ValidationResult(false,
                    "Password must contain at least one lowercase letter");
        }

        if (!DIGIT_PATTERN.matcher(password).find()) {
            return new ValidationResult(false,
                    "Password must contain at least one number");
        }

        if (!SPECIAL_CHAR_PATTERN.matcher(password).find()) {
            return new ValidationResult(false,
                    "Password must contain at least one special character (!@#$%^&*(),.?\":{}|<>)");
        }

        return new ValidationResult(true, null);
    }

    /**
     * Result of password validation
     */
    public static class ValidationResult {
        private final boolean valid;
        private final String message;

        public ValidationResult(boolean valid, String message) {
            this.valid = valid;
            this.message = message;
        }

        public boolean isValid() {
            return valid;
        }

        public String getMessage() {
            return message;
        }
    }
}
